### ISSUE TITLE
SAML2 plugin not working with ADFS 3.0?

### ISSUE THREAD

**Issue by ncalton** [catalyst/moodle-auth_saml2#27]
We have not managed to get the SAML2 plugin working with our ADFS 3.0 implementation running on Windows2012 R2 server. We are currently getting the error: 

Exception - Requester/InvalidNameIDPolicy

**Comment by cristianomariacumer** [comment]
Hi @ncalton,

it'works, we had the same issue but we managed to fix it. Tomorrow I will speak with our ADFS admin to see if he remember what we did to overcome the issue.

**Comment by brendanheywood** [comment]
@dmitriim is this related to your fix?

**Comment by dmitriim** [comment]
@brendanheywood yeah. I believe so. I had the same error.

**Comment by brendanheywood** [comment]
Can you try this and see if it works, if so I'll fold it into master:

```
         'saml:SP',
         'entityID' => "$CFG->wwwroot/auth/saml2/sp/metadata.php",
         'idp' => $saml2auth->config->entityid,
         'NameIDPolicy' => null,  <<<<-----
         'OrganizationName' => array(
             'en' => $SITE->shortname,
         ),
```

**Comment by brendanheywood** [comment]
In the config/authsources.php file

**Comment by ncalton** [comment]
I changed the authsources.php file and the Exception error went away but it still is not working. I get the message that an error has occurred but when I click on the Error details I just get this:
•Activity ID: 00000000-0000-0000-4105-0080000000dc
•Relying party: lmsweb.stfc.ac.uk
•Error time: Thu, 28 Jul 2016 08:10:52 GMT
•Cookie: enabled
•User agent string: Mozilla/5.0 (compatible; MSIE 10.0; Windows NT 6.1; WOW64; Trident/7.0; SLCC2; .NET CLR 2.0.50727; .NET4.0C; .NET4.0E; .NET CLR 3.5.30729; .NET CLR 3.0.30729)

Unfortunately one of our ADFS guys is on holiday and I never had a chance to talk to the other guy today to ask him to check the logs to see if he could get some more information on the error.

**Comment by dmitriim** [comment]
@ncalton looks like IDP error for me. Maybe something is missing in ADFS settings.

**Comment by jlyle29** [comment]
I have an ADFS 2.1 environment and had the same issue. On the ADFS server the following was generated as a user tried to log in.

> The SAML authentication request had a NameID Policy that could not be satisfied. 
Requestor: https://[OUR URL]/auth/saml2/sp/metadata.php 
Name identifier format: urn:oasis:names:tc:SAML:2.0:nameid-format:transient 
SPNameQualifier:  
Exception details: 
MSIS1000: The SAML request contained a NameIDPolicy that was not satisfied by the issued token. Requested NameIDPolicy: AllowCreate: True Format: urn:oasis:names:tc:SAML:2.0:nameid-format:transient SPNameQualifier: . Actual **NameID properties: null.** 

> This request failed. 

> User Action 
Use the AD FS 2.0 Management snap-in to configure the configuration that emits the required name identifier.

After looking over the following two articles I was able to create two claim rules that seem to get things working:
[http://social.technet.microsoft.com/wiki/contents/articles/ad-fs-2-0-survival-guide.aspx#nonMS](url)
[https://social.msdn.microsoft.com/Forums/vstudio/en-US/ea5efcff-4221-4af1-b434-
4be5245cb0fa/nameid-policy-could-not-be-satisfied?forum=Geneva](url)

**Comment by ncalton** [comment]
Thanks for the links - I have forwarded them to the guys in charge of our ADFS (who are all on leave today!).

**Comment by faithlucid** [comment]
hello, was this solved? We have kind of the same issue.

**Comment by brendanheywood** [comment]
I believe this has been solved. If you are having troubles can you dump the actual error you are getting?

**Comment by faithlucid** [comment]
hello, i've opened e new issue here.thanks for reply.

**Comment by cristianocumer** [comment]
I think we should close this issue, AFAIK this plugin is working with ADFS3

**Comment by roperto** [comment]
Closing this as fix in #108 and plugin works on ADFS 3 and 4.
